sda-commons-client-jersey/src/main/java/org/sdase/commons/client/jersey/oidc/OidcClient.java

```diff
@@ -0,0 +1,159 @@
+import org.sdase.commons.client.jersey.oidc.model.OidcState;
+import org.sdase.commons.client.jersey.oidc.rest.OpenIdDiscoveryApi;
+import org.sdase.commons.client.jersey.oidc.rest.model.OpenIdDiscoveryResource;
+import org.sdase.commons.client.jersey.oidc.rest.model.TokenResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Service that can retrieve a new access token. */
+public class OidcClient {
+
+  private static final String CACHE_KEY = "oidcResultKey";
+  private static final long NANOSECOND_CONVERSION_FACTOR = 1000000000;
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(OidcClient.class);
+
+  private final Client client;
+
+  private final OidcConfiguration config;
+
+  private final OpenIdDiscoveryApi discoveryApi;
+
+  private final Cache<String, TokenResource> cache;
+
+  public OidcClient(ClientFactory clientFactory, OidcConfiguration config) {
+    this.config = config;
+    if (config.isDisabled()) {
+      LOGGER.warn("OIDC was disabled.");
+      this.client = null;
+      this.discoveryApi = null;
+      this.cache = null;
+      return;
+    }
+
+    this.client =
+        clientFactory.externalClient(config.getHttpClient()).buildGenericClient("oidc-client");
+
+    this.discoveryApi =
+        createProxy(
+            OpenIdDiscoveryApi.class,
+            WebResourceFactory.newResource(
+                OpenIdDiscoveryApi.class, client.target(config.getIssuerUrl())));
+
+    this.cache =
+        Caffeine.newBuilder()
+            .expireAfter(
+                new Expiry<String, TokenResource>() {
+
+                  @Override
+                  public long expireAfterCreate(
+                      @NonNull String key, @NonNull TokenResource value, long currentTime) {
+                    long expirationTime =
+                        value.getAccessTokenExpiresInSeconds() * NANOSECOND_CONVERSION_FACTOR;
+                    long buffer = 5 * NANOSECOND_CONVERSION_FACTOR;
+                    return (expirationTime - buffer) > 0 ? (expirationTime - buffer) : 0;
+                  }
+
+                  @Override
+                  public long expireAfterUpdate(
+                      @NonNull String key,
+                      @NonNull TokenResource value,
+                      long currentTime,
+                      @NonNegative long currentDuration) {
+                    return currentDuration;
+                  }
+
+                  @Override
+                  public long expireAfterRead(
+                      @NonNull String key,
+                      @NonNull TokenResource value,
+                      long currentTime,
+                      @NonNegative long currentDuration) {
+                    return currentDuration;
+                  }
+                })
+            .build();
+  }
+
+  /**
+   * Retrieves a new access token from the token endpoint using the given {@link OidcConfiguration}.
+   */
+  public synchronized OidcResult createAccessToken() {
```

Comment:
I think it would be nicer for transparency / good design to move the code for the actual retrieval to a separate class.